scintillometry/psrio/psrfits_io.py

```diff
@@ -0,0 +1,383 @@
+        dt = (time - self.header_hdu.start_time).to(u.s)
+        center_off = dt + 1.0 / self.sample_rate * self.samples_per_frame / 2
+        if "OFFS_SUB" in self.columns.names:
+            first_off = self.data['OFFS_SUB'][0]
+            self.data['OFFS_SUB'] = (self.data['OFFS_SUB'] - first_off +
+                                     center_off)
+        else:
+            self._req_columns['OFFS_SUB']['value'] = center_off
+
+    @property
+    def nrow(self):
+        return int(self.header['NAXIS2'])
+
+    @property
+    def nchan(self):
+        return int(self.header['NCHAN'])
+
+    @property
+    def npol(self):
+        return int(self.header['NPOL'])
+
+    @property
+    def nbin(self):
+        return int(self.header['NBIN'])
+
+    @property
+    def samples_per_frame(self):
+        try:
+            return int(self.header['NSBLK'])
+        except:
+            return int(np.prod(self.data_shape)/(self.nrow * self.nchan *
+                                                 self.npol * self.nbin))
+
+    @samples_per_frame.setter
+    def samples_per_frame(self, val):
+        self.header['NSBLK'] = val
+
+    @property
+    def shape(self):
+        raw_shape = self.raw_shape
+        new_shape = namedtuple('shape', ['nsample','nbin', 'nchan', 'npol'])
+        result = new_shape(raw_shape.nrow * raw_shape.samples_per_frame,
+                           raw_shape.nbin, raw_shape.nchan, raw_shape.npol)
+        return result
+
+    @property
+    def raw_shape(self):
+        r_shape = namedtuple('shape', ['nrow', 'samples_per_frame','nbin',
+                                       'nchan','npol'])
+        result = r_shape(self.nrow, self.samples_per_frame, self.nbin,
+                         self.nchan, self.npol)
+        return result
+
+    @property
+    def data_shape(self):
+        # Data are save in the fortran order. Reversed from the header label.
+        d_shape_raw = self.data['DATA'].shape
+        if self.mode == "SEARCH":
+            d_shape_header = (self.nbin, self.nchan, self.npol,
+                              self.samples_per_frame)
+        else:
+            d_shape_header = (self.nbin, self.nchan, self.npol)
+        if d_shape_raw != (self.nrow, ) + d_shape_header[::-1]:
+            raise ValueError("Data shape does not match with the header"
+                             " information")
+        if self.mode == "SEARCH":
+            d_shape = namedtuple('d_shape', ['nrow', 'nsample', 'npol', 'nchan',
+                                             'nbin'])
+            result  = d_shape(self.nrow,  self.samples_per_frame, self.npol,
+                              self.nchan, self.nbin)
+        else:
+            d_shape = namedtuple('d_shape', ['nrow', 'npol', 'nchan', 'nbin'])
+            result  = d_shape(self.nrow,  self.npol, self.nchan, self.nbin)
+        return result
+
+    @property
+    def polarization(self):
+        return self.header['POL_TYPE']
```

Comment:
This has to be an array with the proper shape; seems unlikely to be true by default!

Reply:
It will not work with our pipeline for now. I need to think more about how to convert PSRFITS polarization names to our convention.